in-progress/8025-publishing-tx-objects/design.md

```diff
@@ -17,7 +17,23 @@ Further, we are not requiring validators participating in the pending chain to e
 
 ## Introduction
 
-We first outline the changes needed in various components to support this work.
+### Epochs and committees
+
+We have epochs, which are divided into $E$ slots, which are measured in $S$ seconds.
+
+An epoch has a committee, which is responsible for proposing/validating blocks in the pending chain.
+
+The committee is sampled from a broader validator set based on a seed that gets set when the first proposal in the *previous* epoch is submitted.
+
+The committee changes every epoch.
```

Comment:
TLDR: its cheaper on L1 to do epoch committees, and I don't think the benefit in terms of security is worth the additional cost, but we'll see what the real system shows us!

Right now we have a large validator set that we sample per epoch to form the committee, and then choose random indices within the committee to decide who is the proposer for each slot.

If a committee has a supermajority of dishonest nodes, they could:
- "lie", i.e. include TxObjects in a proposed block that lands on L1 for which a private proof cannot be produced
- include valid TxObjects in a proposed block that lands on L1, but withhold the data needed to verify the proofs (and build the rollup proof)

As far as I know, these are the two major attacks, and both are predicated on a supermajority of dishonest stake in a committee that is attesting to proposals.

In the current design, if a lie/withhold occurs in epoch 1, L1 needs to wait until the end of epoch 1, then $C+P$ slots before it is convinced that the pending chain was bad and reorgs it. So that is a max duration of $E+P+C$ slots that someone could have:
1. Downloaded the bad txobject
2. Executed it to advance the world state
3. Seen something scary (but erroneous) and taken action

Changing the committee per slot I think comes in two flavors:
- Option A: You can still choose an "epoch committee" once per epoch from the validator set, but make it much larger, then choose a "slot committee" for each slot
- Option B: Sample the validator set once per slot. Effectively a special case of option A where the epoch committee and the slot committee are both the entire validator set.

Both of these would presumably reduce the odds that a slot had a committee with supermajority dishonest validators, with option A being cheaper in L1 gas than option B.

But since a committee is being produced which attests to the pending, I don't think either of these options actually change the possible attacks.

Further, I don't think they even change the duration that L1 can be deceived in the event of the "lie" attack, unless the caveat we implement a kind of early warning system. If we had such a "fire alarm", then in either option, if slot 1 had a txobject that was a "lie" *and* they published the proofs to the p2p network, then in slot 2 we could have an honest committee that could warn L1 to not trust the pending chain for the epoch.

I see a few problems with that:
1. That still requires a full L2 slot for the early warning system to get triggered, during which time someone might have taken action elsewhere, particularly if they are a bot/automated.
2. In order for the honest committee to know it was a lie, the dishonest committee needs to publish the bogus private kernel proofs, thereby incriminating themselves. If the honest committee pulls the fire alarm under the withholding case, the dishonest committee could always subsequently reveal it. 
3. It presents another attack vector: a dishonest committee could pull the alarm, even though the epoch is going to get proven, which feels like a scary attack.

Zooming out, we're currently targeting:
- a 1 in a million chance of the committee being supermajority dishonest (based on the assumption that up to 1/3 of the staking asset is malicious).
- 500 L1 gas in overhead per transaction when operating at 10 TPS
- A quantifiable "economic guarantee" that the pending chain can be trusted

So I could see us moving to option A or B if our L1 gas overhead is way under budget, but our odds/security is not where it needs to be.

Reply:
> include valid TxObjects in a proposed block that lands on L1, but withhold the data needed to verify the proofs (and build the rollup proof)

Nit: this requires **all** validators in the committee to be dishonest.

> Both of these would presumably reduce the odds that a slot had a committee with supermajority dishonest validators, with option A being cheaper in L1 gas than option B.

Is there a significant gas cost involved in sampling? Haven't looked at the implementation yet.

> That still requires a full L2 slot for the early warning system to get triggered, during which time someone might have taken action elsewhere, particularly if they are a bot/automated.

True, but being able to inject bogus txs for 32 blocks feels a lot scarier than for a single one. There's another attack we're not considering I think, which is censorship from the validator set. Holding the chain hostage to censorship for an entire epoch is a lot worse than for a single slot.

> In order for the honest committee to know it was a lie, the dishonest committee needs to publish the bogus private kernel proofs, thereby incriminating themselves

How viable would it be to have the epoch and slot committees, for each slot require the signature of the validators from the slot committe only, but have validators only sign if they've validated **all** tx objects in the epoch? This increases the amount of work they have to do, but I think it keeps latency in check?